### ISSUE TITLE
Add support for the HTML5 Notification API

### ISSUE THREAD

**Issue by adam-waldenberg** [omnifaces/omnifaces#508]
Add support for the Notification API.

A sub-component is probably the way to go .. To throw some ideas around,  it would be cool if something like this was doable;

```
<o:messages>
    <o:notification icon="#{bean.notificationImage}"/>
</o:messages>
```
... or this,
```
<o:messages>
    <o:notification icon="#{bean.notificationImage}">
        <o:notificationAction title="Exit" action=#{controller.someAction}>
        <o:notificationAction title="Enter" action=#{controller.someAction}>
    </o:notification>
</o:messages>
```
... or this,
```
<o:messages>
    <o:notification icon="#{bean.notificationImage}" onclick="... " onclose="..."
                    onerror="..." onshow="..."/>
</o:messages>
```
Here are some common use cases:
https://web-push-book.gauntface.com/demos/notification-examples/
https://developer.mozilla.org/en-US/docs/Web/API/Notifications_API/Using_the_Notifications_API

**Comment by melloware** [comment]
@adam-waldenberg So your thought is to intercept FacesMessages for these notifications?  I am wondering if it should just be a complete separate component and not use FacesMessages?

**Comment by adam-waldenberg** [comment]
@melloware Personally, I definitely want something that integrates with FacesMessages.

Doing it as a sub-component would allow to show notifications "the usual way" and at the same time also automatically send them to the notification API. It could also take advantage of what's already implemented in `o:messages`.

However, another way to go about it could be to do a separate component and just revert to a growl-type kind of notification if permissions to the notification API are not given.

**Comment by BalusC** [comment]
Interesting. Will look for 3.4.

**Comment by jepsar** [comment]
@adam-waldenberg I believe notifications are only useful when then [page is not visible](https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API). In that case `FacesMessages` wouldn't be my first choice as they are bound to a request. I think pushing them over a websocket would be more useful.

**Comment by adam-waldenberg** [comment]
@jepsar I dont agree. Websockets dont tie into JSF and the server side all that well and dont work with current code. A normal way to trigger updates on a page on websocket notifications is often something like this;

```<f:socket channel="channel" scope="scope">
    <f:ajax event="event" listener="#{bean.listener}" render="target" />
</f:socket>```

Which would obviously trigger an event on the section that's being updated and also a request - where there are usually FacesMessages being returned on notifications and errors.

There's obviously a use case for supporting both - but for JSF, not supporting FacesMessages doesn't make any sense.

**Comment by jepsar** [comment]
@adam-waldenberg Sure, you could do it like that, but a `FacesMessage` doesn't have all the properties you could use in a browser notification. It will work if you don't care about setting a custom icon per message for example. Another issue will be implementing events. How would you know which message was clicked for example?

I would just take the message object pushed to the socket and process it using JavaScript. There you can determine if the browser is in view. If it is you could update the messages component / growl, if it's not you could send a browser notification.

**Comment by adam-waldenberg** [comment]
@jepsar I already covered an example on how to handle events when I started this issue (though I'm not saying my particular example would be the best one).. It doesnt have to be part of the FacesMessage if its part of the component that handles the message.The same goes for icons. If you want to change the icon, just share a different one via the backing value and let the component fetch the icon on request. Each time a new message arrives you can share a different icon. 

If you take andvatage of normal JSF design principles here - the solution is quite simple-

**Comment by BalusC** [comment]
FYI: I played around it last year but in the end it had a few hard requirements and the event behavior was awkwardly not consistent across browsers. I didn't have the impression I could build a stable JSF component around it.

One of the hard requirements was a service worker which I eventually kicked off with a basic implementation via https://showcase.omnifaces.org/resourcehandlers/PWAResourceHandler introduced in 3.7. This can then easily be expanded to cover this new notification feature. I'll the upcoming year take a second look at the browser state.

**Comment by BalusC** [comment]
Looks like Notifications API has improved significantly since previous week: https://notifications.spec.whatwg.org/ (last update was 15 march 2026). Also the web browser landscape wrt Notifications API is much more uniform and stable, apart from some minor browser-specific quirks such as icon cosmetics.

I think the Notifications API is finally ready to have a stable Faces component, I'll take a third look.

**Comment by BalusC** [comment]
`<o:notification>` has been added. It basically extends `<o:sse>` (also newly introduced in 5.2: https://github.com/omnifaces/omnifaces/issues/935). It doesn't use FacesMessage API as initially requested in ticket as that doesn't make entirely sense. FacesMessage is tied to validation/conversion API. Using HTML notifications for validation/conversion makes no sense.

```
<o:notification channel="notifications" />
```

```
@Inject @Push(type=NOTIFICATION)
private PushContext notifications;

public void send() {
    notifications.send(Notification.createNotificationMessage("Title", "Body message"));
}
```
It reuses o:socket/o:sse Push infrastructure.

**Comment by adam-waldenberg** [comment]
@BalusC Very cool. I think your solution makes perfect sense.
